Portable lockdirs by default

gridbugs opened the pull request:
Enables the portable lockdir feature flag by default. Non-portable lockdirs are still supported, and dune will automatically detect whether a project's lockdir is portable or not and treat it appropriately.

- closes #12694 
- closes #11476

gridbugs committed:
Enable portable lockdirs by default

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs committed:
Update tests to expect portable lockdirs

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs committed:
Stop setting feature flag in portable lockdir tests

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

shonfeder reviewed (commented):
Just had a quick look at the test results. I see some signs of errors that need fixing, IIUC (missing files).

Without some preparatory work to make the tests less noisy, I think it will require quite laborious review to ensure now other errors are sneaking in. Would it be possible to do the former do you think?

gridbugs commented on test/blackbox-tests/test-cases/pkg/compute-checksums-when-missing.t:
```diff
@@ -124,7 +145,15 @@ asserted by the fact that the webserver will only serve the file a single time.
   The source archive will be downloaded from: http://0.0.0.0:2
   Dune will compute its own checksum for this source archive.
-  Solution for dune.lock:
+  Solution for dune.lock
```
When printing a portable solution, the common dependencies are printed in a different list from platform-specific dependencies, like
```
Common deps:
 - a.1.0
 - b.1.0

aarch64 linux deps:
 - c.1.0

aarch64 macos deps:
  - d.1.0
```

Because there are now several lists of dependencies, each preceded by a heading with a colon, if we were to print the "Solution for dune.lock" message with a colon it might look like the heading of an empty list, e.g.
```
Solution for dune.lock:

Common deps:
 - a.1.0
 - b.1.0

aarch64 linux deps:
 - c.1.0

aarch64 macos deps:
  - d.1.0
```
...rather than the heading of the entire output. Let me know if you think there's a clearer way to convey this in message printed by the solver, however I don't think it should block merging this PR.

gridbugs commented:
Updated to normalize the solver output in tests by introducing a helper function `dune_pkg_lock_normalized` which runs `dune pkg lock` and filters the output to be quite close to how it looks with portable lockdirs disabled.

rgrinberg commented on test/blackbox-tests/test-cases/pkg/with-test-dependencies.t:
```diff
@@ -66,8 +67,9 @@ Conflicting packages can't be co-installed:
   $ solve depends-on-foo conflicts-with-foo
-  Error: Unable to solve dependencies for the following lock directories:
-  Lock directory dune.lock:
+  Error:
```
Something weird is happening to this error

rgrinberg reviewed (approved):
Looks OK. Some future work:

1. Remove this experimental flag altogether. The thing to configure is the set of platforms to solve for and not whether the format is enabled or not.

2. When there's a single version of the package, we should drop it from the name. That will make the diffs a lot more readable for upgrades.

gridbugs committed:
Update autolock test to allow rebuilds

Rebuilds happen because the internal package solution generated when
autolocking is specialized to the current platform. This allows building
on platforms which portable lockdirs don't support by default, however
it means that packages built from a pre-existing portable lockdir need
to be rebuilt with the new build plan.

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>
